compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1494,7 +1494,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         found_ty: Ty<'tcx>,
         expr: &hir::Expr<'_>,
     ) {
-        let hir::ExprKind::MethodCall(segment, callee_expr, &[], _) = expr.kind else { return; };
+        let (segment, callee_expr, expr) = match expr.kind {
+            // When `expr` is `foo.clone()`, get `foo` and `clone`.
+            hir::ExprKind::MethodCall(segment, callee_expr, &[], _) => (segment, callee_expr, expr),
+            // When `expr` is `x` in `let x = foo.clone(); x`, get `foo` and `clone`.
+            hir::ExprKind::Path(hir::QPath::Resolved(
+                None,
+                hir::Path { segments: [_], res: crate::Res::Local(binding), .. },
+            )) => {
+                let Some(hir::Node::Pat(
+                    hir::Pat { hir_id, kind: hir::PatKind::Binding(_, _, _, _), .. }
+                )) = self.tcx.hir().find(*binding) else {
+                    return;
+                };
+                let parent = self.tcx.hir().parent_id(*hir_id);
+                let Some(hir::Node::Local(
+                    hir::Local { init: Some(init), .. }
+                )) = self.tcx.hir().find(parent) else {
+                    return;
+                };
```

Comment:
Is there a possibility that this is wrong, considering that `let`s can have arbitrary patterns? e.g. for something like `let (a, b) = x.clone()`

Reply:
Think a reasonable number of these kinds of cases are now handled so resolving but if you think something is missed let me know.

Note that the example given above actually isn't relevant as `(a,b).clone()` will clone `a` and `b` exactly as they are it seems, at least that's what seemed to happen when I tested it. So in our case if `a` and `b` are some `&A` then `(a,b).clone()` will still return a type `(&A, &A)` where our use case is it clones `&A` but cloning `A` would have made types match however `A` doesn't satisfy `Clone` so we hint to say "Maybe add `#[derive(Clone)]` to `A`".